### PR TITLE
JSONRPC L1/L2 Parity tests

### LINKED ISSUE #1162: geth vs zkevm parity tests
- Layer 1: Geth
- Layer 2: ZKEVM Node

As of bcb5f38f97c6fc96658fdc503ffeabeed0d900f3

Test should cover these endpoints:

| `jsonrpc/eth.go` Method Name | Tested? |
|--------------------------------------|----------|
| `BlockNumber()` |  ✅ `468` |
| `Call(arg *txnArgs, number *BlockNumber)` | ✅ `86` |
| `ChainId()` | ✅  done in `Test_Misc` in further commit |
| `EstimateGas(arg *txnArgs, number *BlockNumber)` | ✅ `423` |
| `GasPrice()` | ✅ `239` |
| `GetBalance(address common.Address, number *BlockNumber)` | ✅ `417` |
| `GetBlockByHash(hash common.Hash, fullTx bool)` | ✅ `478` |
| `GetBlockByNumber(number BlockNumber, fullTx bool)` | ✅ `474` |
| `GetCode(address common.Address, number *BlockNumber)` |  ✅ `541` |
| `GetCompilers()` | |
| `GetFilterChanges(filterID argUint64)` |✅ `289` **IS WIP** |
| `GetFilterLogs(filterID argUint64)` | |
| `GetLogs(filter *LogFilter)` | ✅ `378` |
| `GetStorageAt(address common.Address, position common.Hash, number *BlockNumber)` | ✅  done in `Test_Misc` in further commit |
| `GetTransactionByBlockHashAndIndex(hash common.Hash, index Index)` | ✅ `505` |
| `GetTransactionByBlockNumberAndIndex(number *BlockNumber, index Index)` | ✅ `509` |
| `GetTransactionByHash(hash common.Hash)` | ✅ `461` |
| `GetTransactionCount(address common.Address, number *BlockNumber)` |  ✅ `230` |
| `GetBlockTransactionCountByHash(hash common.Hash)` | ✅ `500` |
| `GetBlockTransactionCountByNumber(number *BlockNumber)` | ✅ `488` |
| `GetTransactionReceipt(hash common.Hash)` | ✅ `464` |
| `NewBlockFilter()` | ✅ `279` |
| `NewFilter(filter *LogFilter)` |✅ `299` |
| `NewPendingTransactionFilter(filterID argUint64)` | ✅ `289` |
| `SendRawTransaction(input string)` | ✅ but needs more |
| `UninstallFilter(filterID argUint64)` | ✅ `351` |
| `Syncing()` | ✅  done in `Test_Misc` in further commit |
| `GetUncleByBlockHashAndIndex()` | ❌ method is WIP on L2 |
| `GetUncleByBlockNumberAndIndex()` | ❌ method is WIP on L2 |
| `GetUncleCountByBlockHash()` |❌ method is WIP on L2 |
| `GetUncleCountByBlockNumber()` | ❌ method is WIP on L2 |
| `ProtocolVersion()` | ❌ method doesn't exist on L1 |

Aggregated by functionality with edge cases and creative tests.

### PR DESCRIPTION
**Pull request by KonradIT**
Closes #1162 #1225

### What does this PR do?

Parity tests done using the JSONRPC endpoints in `jsonrpc/eth.go` firing off at L1 and L2. 

**Includes fixed to achieve 1:1 parity status**:

| Discrepancy Issue | Fix |
|----------------------|-----|
| https://github.com/0xPolygonHermez/zkevm-node/issues/1285 |  https://github.com/0xPolygonHermez/zkevm-node/pull/1258/files#diff-72a26f46bee78212d3e0dd9cec954f3002a98abe07a977c3d56da30366f03026R264 |
| https://github.com/0xPolygonHermez/zkevm-node/issues/1283 | https://github.com/0xPolygonHermez/zkevm-node/pull/1258/files#diff-a0622387984e75e762a9865039f3954166b20b8df1ecbd2bf092be9eb3cd5938R24 |
| https://github.com/0xPolygonHermez/zkevm-node/issues/1284 |  https://github.com/0xPolygonHermez/zkevm-node/pull/1258/commits/3135dad3e35be5173ecc3733033c5a5e9ddda69b |

### Reviewers

Main reviewers:

- @arnaubennassar 
- @tclemos 

Codeowner reviewers:

### CODE REVIEW

**Comment by KonradIT** [comment]
@arnaubennassar @tclemos What do you think regarding these possible modifications:

- Use `assert` vs the current `require` so the test won't fail whenever one network mismatches on what's being tested, which would allow us to compare entire L1 vs L2 runs, right now if a test in L2 fails, the next tests will not be executed
- I had to call a test within a test, pretty dirty: https://github.com/0xPolygonHermez/zkevm-node/pull/1258/files#diff-95b37c7fe389e2ecd24846c2f39d65258d467e9b4f443ab78a8799f78e9ec352R350

**Comment by tclemos** [comment]
> @arnaubennassar @tclemos What do you think regarding these possible modifications:
> 
> * Use `assert` vs the current `require` so the test won't fail whenever one network mismatches on what's being tested, which would allow us to compare entire L1 vs L2 runs, right now if a test in L2 fails, the next tests will not be executed
> * I had to call a test within a test, pretty dirty: https://github.com/0xPolygonHermez/zkevm-node/pull/1258/files#diff-95b37c7fe389e2ecd24846c2f39d65258d467e9b4f443ab78a8799f78e9ec352R350

assert package should be used when the rest of the test can be executed regardless of the error that happened.
require package should be used when the rest of the test is not able to execute because of the error that happened.

some naive examples:
require: can't connect to the network
assert: wrong expected value

**Comment by KonradIT** [comment]
TODO: Fix this RPC discrepancy: https://github.com/0xPolygonHermez/zkevm-node/issues/1284

**Review by Psykepro — APPROVED** [review]

File: test/e2e/jsonrpc_test.go [hunk 1]
```diff
@@ -1,186 +1,580 @@
...
+		err = json.Unmarshal(response.Result, &txCount)
+		require.NoError(t, err)
+		require.Equal(t, "0x1", txCount)
+
+		// check if block number is correct
+		count, err := client.TransactionCount(ctx, receipt.BlockHash)
+		require.NoError(t, err)
+		require.Equal(t, uint(1), count)
+
+		// TODO FIXME Investigate discrepancy
```
[the remark this answers is not in the document]
Comment (KonradIT) [reply test/e2e/jsonrpc_test.go:393]: It's this one, unclear how to fix it. https://github.com/0xPolygonHermez/zkevm-node/issues/1284

[the remark this answers is not in the document]
Comment (KonradIT) [reply test/e2e/jsonrpc_test.go]: Testing locally it seems @tclemos 's parallelism fix got rid of it 🚀

[the remark this answers is not in the document]
Comment (tclemos) [reply test/e2e/jsonrpc_test.go]: It was not the parallelism, it was the `defer Teardown()` for each test

**Comment by kind84** [comment]
@KonradIT WDYT about adding a test case here https://github.com/0xPolygonHermez/zkevm-node/blob/58b033f47e11469f965d130b406d38547c170e65/jsonrpc/eth_test.go#L2959  for the new error returned by `NewFilter` ?

**Comment by KonradIT** [comment]
@kind84 which new error?

**Comment by kind84** [comment]
> @kind84 which new error?

@KonradIT even though it is covered by e2e test , IMHO [the new check](https://github.com/0xPolygonHermez/zkevm-node/blob/25d6e325cf7d12bc144bca133b9337a125e84fad/jsonrpc/eth.go#L579) would deserve a matching test case in the UT [here](https://github.com/0xPolygonHermez/zkevm-node/blob/58b033f47e11469f965d130b406d38547c170e65/jsonrpc/eth_test.go#L2959). WDYT?

File: jsonrpc/eth.go [hunk 2]
```diff
@@ -576,11 +576,10 @@ func (e *Eth) NewBlockFilter() (interface{}, rpcError) {
 // to notify when the state changes (logs). To check if the state
 // has changed, call eth_getFilterChanges.
 func (e *Eth) NewFilter(filter *LogFilter) (interface{}, rpcError) {
-	if filter.BlockHash != nil && (filter.FromBlock != nil || filter.ToBlock != nil) {
-		return rpcErrorResponse(invalidParamsErrorCode, "invalid argument 0: cannot specify both BlockHash and FromBlock/ToBlock, choose one or the other", nil)
-	}
 	id, err := e.storage.NewLogFilter(*filter)
-	if err != nil {
+	if errors.Is(err, ErrFilterInvalidPayload) {
+		return rpcErrorResponse(invalidParamsErrorCode, "invalid argument 0: cannot specify both BlockHash and FromBlock/ToBlock, choose one or the other", nil)
```
Comment (kind84) [inline jsonrpc/eth.go:581]: @KonradIT I think that probably here you can do this now:
```go
return rpcErrorResponse(invalidParamsErrorCode, err.Error(), nil)
```
It should take the error message from the `ErrFilterInvalidPayload` in `pgstorage.go`

**Review by kind84 — APPROVED** [review]
LGTM :clap:
